lib/libmuser.c

```diff
@@ -663,29 +664,33 @@ lm_get_region(loff_t pos, size_t count, loff_t *off)
     return r;
 }
 
+static uint32_t
+region_size(lm_ctx_t *lm_ctx, int region)
+{
+        assert(region >= LM_DEV_BAR0_REG_IDX && region <= LM_DEV_VGA_REG_IDX);
+        return lm_ctx->pci_info.reg_info[region].size;
+}
+
+static uint32_t
+pci_config_space_size(lm_ctx_t *lm_ctx)
+{
+    return region_size(lm_ctx, LM_DEV_CFG_REG_IDX);
+}
+
 static ssize_t
 handle_pci_config_space_access(lm_ctx_t *lm_ctx, char *buf, size_t count,
-                               loff_t pos, bool is_write,
-                               lm_region_access_t *pci_config_fn)
+                               loff_t pos, bool is_write)
 {
-    int r1, r2 = 0;
+    int ret;
 
-    r1 = cap_maybe_access(lm_ctx->caps, lm_ctx->pvt, buf, count, pos, is_write);
-    if (r1 < 0) {
+    count = MIN(pci_config_space_size(lm_ctx), count);
+    ret = cap_maybe_access(lm_ctx->caps, lm_ctx->pvt, buf, count, pos, is_write);
+    if (ret < 0) {
         lm_log(lm_ctx, LM_ERR, "bad access to capabilities %u@%#x\n", count,
                pos);
-        return r1;
-    }
-    buf += r1;
```

Comment:
It looks like after this change, pci_config_fn for LM_DEV_CFG_REG_IDX will never be called even if it falls in PCI_STD_HEADER_SIZEOF?

Reply:
`pci_config_fn` should never be called for `PCI_STD_HEADER_SIZEOF`. Now that you raised this point, the question is whether there should be a user callback for accesses past the header and the capabilities. I seem to remember that we should handle all accesses to the PCI cofiguration space but I might be wrong. @franciozzy ?

Reply:
I don't think there should be any callbacks for the config space besides vendor capabilities. We originally tried to make it so config accesses that were not handled by libmuser would fallback somehow, but I don't think that was a good decision. Libmuser handle everything there.